rep-2001.rst

```diff
@@ -763,14 +764,24 @@ Simulation
       extends:  [ros_base]
       packages: [ros_ign_bridge, ros_ign_gazebo, ros_ign_image, ros_ign_interfaces]
 
+Dev Tools
+"""""""""
+
+::
+
+  - dev_tools:
+      packages: [python3-colcon-common-extensions, python3-flake8, python3-pip,
+                 python3-pytest, python3-pytest-cov, python3-rosdep, python3-setuptools,
+                 python3-vcstool, cmake, git, wget]
```

Comment:
The reason I've been hesitant on this is because this new package is a different kind of beast from all of the other ones in here.  In particular, all of the rest of the variants here depend on things directly generated by the ROS buildfarm (though of course they indirectly depend on things from the underlying operating system).

This new variant, instead, directly depends on a combination of things from the underlying operating system (wget, git, etc), as well as packages that are imported into ROS from elsewhere (python3-colcon-common-extensions).

That doesn't necessarily mean we shouldn't do it, but it is giving me pause both from a precedent standpoint (should we allow other collections of packages?), as well as a practical standpoint (will the buildfarm happily build this?).  I've been talking to @nuclearsandwich and @cottsay a little bit more about this, they may leave their thoughts about this as well.

Reply:
(That's not a threat, I more mean that I'm not adamant that it must be part of the REP and needs to go in `variants`)

Reply:
Cementing some of my thoughts from the discussion Chris alluded to, I think that the biggest outward concern for this metapackage is that by using a ROS package for the variant, we either _only_ release this variant in a single distribution like Rolling or we end up with duplicate variants: `ros-rolling-dev-tools`, `ros-humble-dev-tools`, `ros-iron-dev-tools` which all install the same system-wide packages. This will generate some confusion as users ask questions like "Do I need the dev-tools package for a specific distribution?" or "do I need to source the distribution setup script to use these dev tools?"

Implementing this metapackage as an in-ROS variant is the expedient way to make it happen if those concerns don't worry anyone else. Overall I think having a package like this would greatly improve and streamline the setup instructions for working with source distributions of ROS on supported platforms and in that same discussion we talked about possibly having a `ros-build-essential` package which builds on the platform compiler toolchain to codify and provide the baseline packages which should be present during build farm package builds implicitly.

The rosdistro independent method for creating `ros-build-essential` and `ros-dev-tools` metapackages would be to create them with `equivs` and comparable tools for RPM distributions and distribute them through the ROS bootstrap repositories.

Reply:
I get the concern of having duplicate variants, but I worry about asserting that the dependencies will be the same for all distributions forever and always. What if the ghost of Dirk Thomas comes back and writes `colcon2` for ROS-2-K? If you install `ros-build-essential` does it have `python3-colcon-common-extensions` or `python3-colcon2`?

Reply:
Is `equivs` a specific repo like `variants`?

Reply:
> I get the concern of having duplicate variants, but I worry about asserting that the dependencies will be the same for all distributions forever and always.

Yeah, that has crossed my mind as well but as it stands all of the tools in the variant now are ROS independent. For a time ROS 2's ament_package and ament_tools were included in each distribution but both were eventually dropped in favor of independent tools (ament_package is technically still in Rolling but it contains _only_ ROS 2 specific templates and no actual code).

It does not make sense, in my opinion, for this metapackage to include build tools (such as cmake, catkin, ament_cmake) since each package is expected to explicitly depend on the build tool that it uses, but if build tools were included, that would add weight to the ROS-dependent approach.

> What if the ghost of Dirk Thomas comes back and writes `colcon2` for ROS-2-K? If you install `ros-build-essential` does it have `python3-colcon-common-extensions` or `python3-colcon2`?

In point of fact, I think `ros-build-essential` has neither because neither is critical to build a ROS package on the build farm. But `ros-dev-tools` would either include both tools or include only the one that is the recommended build tool for all current distributions as the current proposal does by not including catkin-tools. For bootstrapping reasons, I don't foresee a situation where the officially recommended build tool is ever a ROS-dependent package like ament_tools was again.

> Is `equivs` a specific repo like `variants`?

Whoops, sorry for the jargon. [`equivs`](https://manpages.debian.org/testing/equivs/equivs-build.1.en.html) is a tool for building small Debian packages using just a control file instead of an entire package and `debian/` which makes creating small metapackages really straightforward (assuming you're only supporting Debian-derived platforms). It has been used in ROS 2 to create [trivial packages that satisfy package dependencies](https://github.com/nuclearsandwich/ros2-equivs) on platforms where they are not available since the package.xml format has no way to describe optional/opportunistic dependencies.

Reply:
NB: I'm explaining my rationale, but I'll be happy with whatever solution you and other experts come up with. 

All I want at the end of the day is a workflow where I can do this.

    sudo apt install ros-$ROSDISTRO-ros-base
    sudo apt install ros-SOMETHING
    rosdep install --from-paths src --ignore-src -y
    colcon build

Reply:
> All I want at the end of the day is a workflow where I can do this.

Based on the upbeat tempo of the offline discussion we were having I think you have @clalancette, @cottsay, and myself all convinced that this would be a big improvement.

I think that we've passed the phase where we're resistant to the proposal (which for me was based on the novelty and not some principled opposition) and into the phase where we're trying to think through the implementation options and the effects of each of them to arrive at an optimal long term solution.